### PR TITLE
Add Image Analysis feature to contact page; upon submitting comment w…

### PR DESCRIPTION
**Pull request by nonsensicle**
…ith image, the server uses Vision API to return image labels/confidences and detected object names/confidences. To do: work on CSS formatting & figure out how one would store bounding boxes in Datastore for box drawing around objects.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -195,7 +251,92 @@ private String getImageKey(HttpServletRequest request, String formElementID) {
...
+    long currByteIndex = 0;
+    boolean continueRead = true;
+    while(continueRead) {
+      // Fetch a portion of the image bytes from Blobstore.
+      byte[] temp = bs.fetchData(imageKey, currByteIndex, currByteIndex + fetchSize - 1); 
+      outputBytes.write(temp);
+
+      // If fewer bytes than requested are read, then the end was reached.
+      if (temp.length < fetchSize) {
+        continueRead = false;
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I don't think this is necessary - you can just break here.

Actually a do-while loop might be cleaner here, as you can check the looping condition temp.length < fetchSize at end of the loop.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -195,7 +251,92 @@ private String getImageKey(HttpServletRequest request, String formElementID) {
...
+    // Use a Feature to make an annotate-image request; add this request to a list of such.
+    Feature feature = Feature.newBuilder().setType(Feature.Type.OBJECT_LOCALIZATION).build();
+    AnnotateImageRequest request = 
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    // Batch annotate all the image requests in the list (only one, here) and get the responses.
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Looks like batchAnnotateImages can throw ApiException (which is a runtime exception - meaning your compiler won't complain if you are not catching it). If batchAnnotateImages throws ApiException, client.close would never get called. You'd need a try-catch (or try-with-resource) block here.

Generally speaking, we should catch any exception third-party api throws and process them as close to the source of the call as possible.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -117,15 +138,50 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    if (imageKey != null) { 
+      // Create string from key to store in datastore.
+      imKey = imageKey.getKeyString(); 
+
+      byte[] imageBytes = getBlobBytes(imageKey);
+      List<LocalizedObjectAnnotation> imageObjects = getImageObjects(imageBytes);
+      List<EntityAnnotation> imageLabels = getImageLabels(imageBytes);
+      
+      // Build the lists of image descriptions and scores.
+      for (EntityAnnotation label: imageLabels) {
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Both imageLabels and imageObjects can be `null` - this will cause an NullPointerException.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -31,6 +31,18 @@
...
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.cloud.vision.v1.LocalizedObjectAnnotation;
+import com.google.cloud.vision.v1.NormalizedVertex;
+import com.google.cloud.vision.v1.*;
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Generally we prefer not importing .* and list everything you want to import individually.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -195,7 +251,92 @@ private String getImageKey(HttpServletRequest request, String formElementID) {
...
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    // Batch annotate all the image requests in the list (only one, here) and get the responses.
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();    // client should always be closed
+    AnnotateImageResponse imageResponse = batchResponse.getResponsesList().get(0);
+ 
+    if (imageResponse.hasError()) {
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Do you have to return `null` here? Look like having this set doesn't [necessarily mean there's no annotation](https://googleapis.dev/java/google-cloud-vision/latest/com/google/cloud/vision/v1/AnnotateImageResponse.html#hasError--).

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -76,16 +78,73 @@ function createCommentElement(fname, surname, date, message, blobKey) {
...
+          headerLabels = document.createElement("tr");
+          headerLabels.innerHTML = "<th>Image Label</th><th>Confidence</th>"
+          tableHeader.appendChild(headerLabels);
+          table.appendChild(tableHeader);
+ 
+          // Building body rows. 
+          tableBody = document.createElement("tbody");
+          for (i = 0; i < imageDescs.length; i++) {
+            row = document.createElement("tr");
+            row.innerHTML = "<td>" + imageDescs[i] + "</td><td>" + descScores[i] + "</td>";
```
Comment (hanpeiz) [inline portfolio/src/main/webapp/script.js]: Should format this number a bit better - having 10+ decimals isn't really relevant to most visitors.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (nonsensicle) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think it's worth returning null since it's not exactly clear what the annotation looks like in this case; this also makes it easier to test up above if there are image labels or not, just test == null.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (nonsensicle) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I'm not sure if I'm misunderstanding how image bytes work, but does this old construction actually cut off a very small amount of the image data at the end?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -297,17 +301,21 @@ private BlobKey getImageKey(HttpServletRequest request, String formElementID) {
     requests.add(request);
 
     // Batch annotate all the image requests in the list (only one, here) and get the responses.
-    ImageAnnotatorClient client = ImageAnnotatorClient.create();
-    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
-    client.close();    // client should always be closed
-    AnnotateImageResponse imageResponse = batchResponse.getResponsesList().get(0);
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+      client.close();    // client should always be closed
```
Comment (hanpeiz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This is no longer needed for in a try-with-resource statement. If you have a AutoCloseable object (which ImageAnnotatorClient is) defined inside parenthesis after try(), it will automatically call .close() whether an exception has occurred or not.

File: woodchip-step-2020-f63454181d8d.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (hanpeiz) [inline woodchip-step-2020-f63454181d8d.json]: This shouldn't be here, I think? Something went wrong with .gitignore?
